helios-crtauth/src/main/java/com/spotify/helios/auth/crt/CrtAuthenticationPlugin.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2014 Spotify AB.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.helios.auth.crt;
+
+import com.google.auto.service.AutoService;
+import com.google.common.annotations.VisibleForTesting;
+
+import com.spotify.crtauth.CrtAuthServer;
+import com.spotify.crtauth.keyprovider.KeyProvider;
+import com.spotify.helios.auth.AuthenticationPlugin;
+
+import org.springframework.ldap.core.LdapTemplate;
+import org.springframework.ldap.core.support.LdapContextSource;
+
+import java.util.Map;
+
+@AutoService(AuthenticationPlugin.class)
+public class CrtAuthenticationPlugin implements AuthenticationPlugin<CrtAccessToken> {
+
+  private final Map<String, String> environment;
+
+  public CrtAuthenticationPlugin() {
+    this(System.getenv());
+  }
+
+  @VisibleForTesting
+  protected CrtAuthenticationPlugin(Map<String, String> environment) {
+    this.environment = environment;
+  }
+
+  @Override
+  public String schemeName() {
+    return "crtauth";
+  }
+
+  @Override
+  public ServerAuthentication<CrtAccessToken> serverAuthentication() {
+    // only validate the presence of environment variables when this method is called, as opposed to
+    // in the constructor, as the client-side code will not use the same environment variables
+    final String ldapUrl =  getRequiredEnv("CRTAUTH_LDAP_URL");
+    final String ldapSearchPath = getRequiredEnv("CRTAUTH_LDAP_SEARCH_PATH");
+    final String serverName = getRequiredEnv("CRTAUTH_SERVERNAME");
+    final String secret = getRequiredEnv("CRTAUTH_SECRET");
+    final String ldapFieldNameOfKey = getOptionalEnv("CRTAUTH_LDAP_KEY_FIELDNAME", "sshPublicKey");
+    final int tokenLifetimeSecs = getOptionalEnv("CRTAUTH_TOKEN_LIFETIME_SECS", 540);
+
+    final LdapContextSource contextSource = new LdapContextSource();
+    contextSource.setUrl(ldapUrl);
+    contextSource.setAnonymousReadOnly(true);
+    contextSource.setCacheEnvironmentProperties(false);
+
+    final LdapTemplate ldapTemplate = new LdapTemplate(contextSource);
+
+    // TODO (mbrown): this should be general, support reading keys from flat files etc
```

Comment:
Should we allow combining multiple types of KeyProviders? Users might want to have LDAP for humans but flat files for CI agents because their IT department doesn't like adding non-human LDAP credentials. I'm not sure if this would make the implementation less secure though, eg timing attacks.